js/modules/k6/grpc/client.go

```diff
@@ -0,0 +1,479 @@
+	grpcstats "google.golang.org/grpc/stats"
+	"google.golang.org/grpc/status"
+
+	"github.com/loadimpact/k6/js/common"
+	"github.com/loadimpact/k6/lib"
+	"github.com/loadimpact/k6/lib/metrics"
+	"github.com/loadimpact/k6/stats"
+)
+
+//nolint: lll
+var (
+	errInvokeRPCInInitContext = common.NewInitContextError("invoking RPC methods in the init context is not supported")
+	errConnectInInitContext   = common.NewInitContextError("connecting to a gRPC server in the init context is not supported")
+)
+
+// Client reprecents a gRPC client that can be used to make RPC requests
+type Client struct {
+	mds  map[string]*desc.MethodDescriptor
+	tags map[string]string
+	conn *grpc.ClientConn
+}
+
+// NewClient creates a new gPRC client to make invoke RPC methods.
+func (*GRPC) NewClient(ctxPtr *context.Context) interface{} {
+	rt := common.GetRuntime(*ctxPtr)
+
+	return common.Bind(rt, &Client{}, ctxPtr)
+}
+
+// MethodInfo holds ifromation on any parsed method descriptors that can be used by the goja VM
+type MethodInfo struct {
+	FullMethod     string
+	IsClientStream bool
+	IsServerStream bool
+}
+
+// Response is a gRPC response that can be used by the goja VM
+type Response struct {
+	Status   codes.Code
+	Message  goja.Value
+	Headers  map[string][]string
+	Trailers map[string][]string
+	Error    goja.Value
+}
+
+// Load will parse the given proto files and make the file descriptors available to request.
+func (c *Client) Load(ctxPtr *context.Context, importPaths []string, filenames ...string) ([]MethodInfo, error) {
+	if lib.GetState(*ctxPtr) != nil {
+		return nil, errors.New("load must be called in the init context")
+	}
+
+	parser := protoparse.Parser{
+		ImportPaths:      importPaths,
+		InferImportPaths: len(importPaths) == 0,
+	}
+
+	fds, err := parser.ParseFiles(filenames...)
+	if err != nil {
+		return nil, err
+	}
+
+	var rtn []MethodInfo
+	c.mds = make(map[string]*desc.MethodDescriptor)
+	for _, fd := range fds {
+		for _, sd := range fd.GetServices() {
+			for _, md := range sd.GetMethods() {
+				var s strings.Builder
+				s.WriteString(sd.GetFullyQualifiedName())
+				s.WriteRune('/')
+				s.WriteString(md.GetName())
+				name := s.String()
+				c.mds[name] = md
+				rtn = append(rtn, MethodInfo{
+					FullMethod:     name,
+					IsClientStream: md.IsClientStreaming(),
+					IsServerStream: md.IsServerStreaming(),
+				})
+			}
+		}
+	}
+
+	return rtn, nil
+}
+
+type transportCreds struct {
+	credentials.TransportCredentials
+	errc chan<- error
+}
+
+func (t transportCreds) ClientHandshake(ctx context.Context,
+	addr string, in net.Conn) (net.Conn, credentials.AuthInfo, error) {
+	out, auth, err := t.TransportCredentials.ClientHandshake(ctx, addr, in)
+	if err != nil {
+		t.errc <- err
+	}
+
+	return out, auth, err
+}
+
+// Connect is a block dial to the gRPC server at the given address (host:port)
+func (c *Client) Connect(ctxPtr *context.Context, addr string, params map[string]interface{}) (bool, error) {
```

Comment:
This is me being unfamiliar with the goja bridge: when just returning the error, I (and users) would need to explicitly check the error in the js code, but if I returned a `(value, error)` then the error would be "thrown" and give better error feedback to the user on connection errors (vs failing later because the connection is `nil`).

Any advice here on what is best?

Reply:
hmm I'll have to check this and get back to you, I don't remember all of the details on how the go <-> js error handling worked

Reply:
From what I can grok of the bridge black magic, you are unfortunately correct. We probably should refactor the bridge logic, but that definitely won't happen any time soon, so resolving this for now. It's not unlikely that `client.Connect()` will have to return some sensible result eventually, anyway, so we just will treat this as a `void` method for now, for the purposes of the docs, i.e. it either throws an error or is successful and doesn't return anything.

Reply:
Would it be better to return `interface{}` rather than `bool` then?

Reply:
Hmm maybe that would be a bit better, but I don't think it really matters - we don't know if this will ever return anything useful in the future. But `interface{}` will allow you to `return nil`, so might be a bit cleaner if some linter doesn't complain and goja and the bridge are fine with it.

In any case, a comment to explain _why) we're returning 2 parameters would be the most useful change. And the important part is that we don't want to have to make a breaking change in the future, which should be covered by us simply not documenting that this function returns anything, I think.